diffkemp/simpll/DifferentialFunctionComparator.cpp

```diff
@@ -21,67 +22,100 @@
-                                    idxMD->getOperand(0))->getValue());
-
-                    // Get value of the old offset from the current index
-                    auto oldOffset = DL.getStructLayout(
-                            TypeItL.getStructType())->getElementOffset(
-                            dyn_cast<ConstantInt>(ix)->getZExtValue());
-                    // Get value of the new offset (from metadata)
-                    auto newOffset = DL.getStructLayout(
-                            TypeItR.getStructType())->getElementOffset(
-                            ixAlign->getValue().getZExtValue());
-                    OffsetL += APInt(BitWidth, newOffset - oldOffset, true);
-                }
-            }
-        }
-        return cmpAPInts(OffsetL, OffsetR);
-    }
-    if (int Res = cmpTypes(GEPL->getSourceElementType(),
-                           GEPR->getSourceElementType()))
+    if (int Res = cmpNumbers(GEPL->getNumIndices(), GEPR->getNumIndices()))
         return Res;
 
-    if (int Res = cmpNumbers(GEPL->getNumOperands(), GEPR->getNumOperands()))
-        return Res;
+    bool allIndicesSame = true;
 
-    for (unsigned i = 0, e = GEPL->getNumOperands(); i != e; ++i) {
-        if (int Res = cmpValues(GEPL->getOperand(i), GEPR->getOperand(i)))
-            return Res;
-    }
+    if (GEPL->hasAllConstantIndices() && GEPR->hasAllConstantIndices()) {
+        std::vector<Value *> IndicesL;
+        std::vector<Value *> IndicesR;
+
+        const GetElementPtrInst *GEPiL = dyn_cast<GetElementPtrInst>(GEPL);
+        const GetElementPtrInst *GEPiR = dyn_cast<GetElementPtrInst>(GEPR);
+
+        if (!GEPiL || !GEPiR)
+            return OldFunctionResult;
+
+        for (auto idxL = GEPL->idx_begin(), idxR = GEPR->idx_begin();
+         idxL != GEPL->idx_end() && idxR != GEPR->idx_end();
+         ++idxL, ++idxR) {
+            auto ValueTypeL = GEPiL->getIndexedType(GEPiL->getSourceElementType(),
+                                                    ArrayRef<Value *>(
+                                                            IndicesL));
 
-    return 0;
+            auto ValueTypeR = GEPiR->getIndexedType(GEPiR->getSourceElementType(),
+                                                    ArrayRef<Value *>(
+                                                            IndicesR));
+
+            auto NumericIndexL = dyn_cast<ConstantInt>(idxL->get())->getValue();
+            auto NumericIndexR = dyn_cast<ConstantInt>(idxR->get())->getValue();
+
+            if (!ValueTypeL->isStructTy() || !ValueTypeR->isStructTy()) {
+                // If the indexed type is not a structure type, the indices have
+                // to match in order for the instructions to be equivalent
+                if (!cmpValues(idxL->get(), idxR->get()))
+                    allIndicesSame = false;
+
+                IndicesL.push_back(*idxL);
+                IndicesR.push_back(*idxR);
+                break;
```

Comment:
Do we really want to break if types are not of structure type but cmpValues returns 0? What about the following indices, don't we want to compare those as well?